src/mbedtls_stream.c

```diff
@@ -0,0 +1,474 @@
+#define X509_CERT_DIR_EVP        "SSL_CERT_DIR"
+#define X509_CERT_FILE_EVP       "SSL_CERT_FILE"
+
+mbedtls_ssl_config *git__ssl_conf;
+mbedtls_entropy_context *mbedtls_entropy;
+
+#define GIT_SSL_DEFAULT_CIPHERS "ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-GCM-SHA256:DHE-DSS-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-DSS-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA:DHE-DSS-AES128-SHA256:DHE-DSS-AES256-SHA256:DHE-DSS-AES128-SHA:DHE-DSS-AES256-SHA:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA"
+
+/**
+ * This function aims to clean-up the SSL context which
+ * we allocated.
+ */
+static void shutdown_ssl(void)
+{
+    if (git__ssl_conf) {
+        mbedtls_x509_crt_free(git__ssl_conf->ca_chain);
+        git__free(git__ssl_conf->ca_chain);
+        mbedtls_ctr_drbg_free(git__ssl_conf->p_rng);
+        git__free(git__ssl_conf->p_rng);
+        mbedtls_ssl_config_free(git__ssl_conf);
+        git__free(git__ssl_conf);
+        git__ssl_conf = NULL;
+    }
+    if (mbedtls_entropy) {
+        mbedtls_entropy_free(mbedtls_entropy);
+        git__free(mbedtls_entropy);
+        mbedtls_entropy = NULL;
+    }
+}
+
+int git_mbedtls_stream_global_init(void)
+{
+    int ret, isdir;
+    char *crtpath;
+    struct stat statbuf;
+
+    mbedtls_ctr_drbg_context *ctr_drbg;
+
+    mbedtls_entropy = git__malloc(sizeof(mbedtls_entropy_context));
+    mbedtls_entropy_init(mbedtls_entropy);
+
+    // Seeding the random number generator
+    ctr_drbg = git__malloc(sizeof(mbedtls_ctr_drbg_context));
+    mbedtls_ctr_drbg_init(ctr_drbg);
+    if (mbedtls_ctr_drbg_seed(ctr_drbg,
+                mbedtls_entropy_func,
+                mbedtls_entropy, NULL, 0) != 0) {
+        mbedtls_ctr_drbg_free(ctr_drbg);
+        mbedtls_entropy_free(mbedtls_entropy);
+        git__free(ctr_drbg);
+        git__free(mbedtls_entropy);
+        return -1;
+    }
+
+    // configure TLSv1
+    git__ssl_conf = git__malloc(sizeof(mbedtls_ssl_config));
+    mbedtls_ssl_config_init(git__ssl_conf);
+    if ( mbedtls_ssl_config_defaults(git__ssl_conf,
+                MBEDTLS_SSL_IS_CLIENT,
+                MBEDTLS_SSL_TRANSPORT_STREAM,
+                MBEDTLS_SSL_PRESET_DEFAULT ) != 0) {
+        mbedtls_ctr_drbg_free(ctr_drbg);
+        git__free(ctr_drbg);
+        mbedtls_ssl_config_free(git__ssl_conf);
+        git__free(git__ssl_conf);
+        git__ssl_conf = NULL;
+        return -1;
+    }
+
+    mbedtls_ssl_conf_authmode(git__ssl_conf, MBEDTLS_SSL_VERIFY_REQUIRED);
+    mbedtls_ssl_conf_rng(git__ssl_conf, mbedtls_ctr_drbg_random, ctr_drbg);
+
+    // find locations for which CA certificates
```

Comment:
libgit2 is very much not in the business of guessing where SSL certificates live on everyone's machine. A distribution-provided mbedtls should know already. And if this is being shipped as part of a bundle, the application is the only one which would know what to use.